### PR TITLE
simplify configuration of probe bed postions

### PR DESCRIPTION
**Pull request by fiveangle**
Auto-configure maximized probe bed positions if user has not manually set them.

### CODE REVIEW

**Comment by fiveangle** [comment]
In response to continued user questions like #7222

**Comment by Tannoo** [comment]
I'm not complaining, but this has been attempted before: #5581

**Comment by fiveangle** [comment]
Yes, and the author failed to submit to rcbugfix then abndoned the idea.  I had meant to submit it long before #5581, so when I saw yet another user confused by the config, I thought it time to resolve.

I'm not sure what Scott meant by "...take into consideration probe offsets" because I believe this does (unless there's a different probe offset other than that in Configuration.h).  If this doens't take into consideration probe offsets, I'm all for having it pointed out what was meant.

**Comment by Tannoo** [comment]
I agree and have the same question.

**Comment by fiveangle** [comment]
But thanks for pointing out #5581.  I didn't look for it when I wrote this PR, and looking at it, the probe offsets comments appear related to the earlier submission where the author had a very rudimentary solution.  However, Scott did want each element evaluated independently, so I'll add those parens as he requested...)

**Comment by benlye** [comment]
I think this is a great idea, and now that it's been done, glaringly obvious - if Marlin is smart enough to tell the user that the probe boundaries are wrong, why can't it automatically derive the correct values?

A lot of people new to Marlin seen to struggle with the maths/logic to do this, and I've helped a number of people with working out what these values ought to be for their geometry.  Doing it automatically eliminates one more source of confusion and lowers the barrier to getting up and running.

**Comment by Tannoo** [comment]
Oh, it can. Marlin is so widely used now with users of all abilities, that it should, in my opinion, be as automated as possible with the options to go partially or fully manual.

Like a modern SLR camera.

File: Marlin/Configuration.h [hunk 1]
```diff
@@ -852,11 +852,13 @@
...
-  #define RIGHT_PROBE_BED_POSITION 170
-  #define FRONT_PROBE_BED_POSITION 20
-  #define BACK_PROBE_BED_POSITION 170
+  // Boundaries for probing (where the probe can reach) are automatically
+  // maximized based on bed size, probe offset, and minimum probe edge.
+  // Uncomment here to set manually.
+  //#define LEFT_PROBE_BED_POSITION 15
+  //#define RIGHT_PROBE_BED_POSITION 170
+  //#define FRONT_PROBE_BED_POSITION 20
+  //#define BACK_PROBE_BED_POSITION 170
```
Comment (FHeilmann) [inline Marlin/Configuration.h:862]: I would suggest that you leave the original configuration.h untouched (so uncomment those lines) and just keep the updated instructions. People are used to setting this themselves and I would leave new functionality optional.

Multiline comments are better with `/** ... **/` and make sure that you update the example configurations as well.

**Comment by FHeilmann** [comment]
Following up on my comment, and some justification on why I suggest keeping this optional, for now:

On my printer, for example, the bed is not perfectly centered w.r.t. the endpoints of the X and Y axis, so I need to use different distances for the FRONT and BACK values. This requires me to manually configure the offsets, because your approach assumes perfect symmetry of the bed w.r.t. to the X and Y axis limits.

I am certain that there are a lot of printers out there with similar situations, so keeping the manual positions enabled by default, while giving the automatic computation as an option seems to be the safer route to me. Manual is sure to work for everyone (while ofc being more tedious), while automatic has some implications and assumptions on printer symmetry and thus sacrifices "generality"

**Comment by Tannoo** [comment]
That can also be reversed..

Most printers are likely symmetrical so this would be fine. Those that aren't need to be manually set.

**Comment by FHeilmann** [comment]
However, manual setup would work for both symmetrical and asymmetrical ones, while automatic only works for symmetrical ones.

Please note that I am just brainstorming (and not criticizing) here, I personally don't mind if the change is merged like this, as I know pretty well how to set up my printers.

Further I would argue that most printers out there these days are Prusa style printers and (from what I've seen) most of those are, at best, symmetrical on the X-Axis, but not on the Y-Axis.

**Comment by Tannoo** [comment]
```
X_MIN_POS
X_MAX_POS
Y_MIN_POS
Y_MAX_POS
```
If those are set, wouldn't any calulation of those be good? Even asymmetrical beds?

**Comment by FHeilmann** [comment]
In my mind that would be more complicated than just setting the probe edges by hand. If you have, for example, a Y-Axis where the nozzle can significantly go outside of the bed only on one edge, you would have to set the maxima and minima such that the nozzle cannot go outside of the bed (or use a large MESH INSET, which results in sacrificing probeable areas on the other end of the bed). However, moving the nozzle outside of the bed may still be desired for things such as prime or nozzle clean movements.

Ultimately, IMHO both of our arguments are valid depending on which philosophy is preferred and the outcome will depend on our [BDFL](https://en.wikipedia.org/wiki/Benevolent_dictator_for_life) @thinkyhead :smile:

**Comment by Tannoo** [comment]
>In my mind that would be more complicated than just setting the probe edges by hand.

Well, the bed size and the probe offsets are put in by hand. Can't the probe boundaries be figured out from there?

Sanity check file says yes.

**Comment by benlye** [comment]
> Well, the bed size and the probe offsets are put in by hand. Can't the probe offsets be figured out from there?

Do we have bed size defined anywhere explicitly, or is it inferred from the travel limits?  

Is this solvable by defining something like `BUILD_AREA_SIZE_X` and `BUILD_AREA_SIZE_Y` and making the probe limits a function of build area, probe offset, and `MIN_PROBE_EDGE`, thus separating the bed limits from travel limits?

**Comment by FHeilmann** [comment]
I think the key here is to specify (for cartesians) the offsets of the bed corner closest to the `X0 Y0` position to said `X0 Y0` position, the bed size and the inset. From there, one would be able to compute a symmetric probing plane across the bed w.r.t. the edges.

For example: In my (admittedly very special case), ignoring clamping to avoid hitting limit switches and adjustments due to nozzle-probe offsets:

```
BED_SIZE_X = 305
BED_SIZE_Y = 305
BED_CORNER_OFFSET_X = 12 (nozzle outside of bed on the left side)
BED_CORNER_OFFSET_Y = -6 (nozzle cannot physically go outside the bed towards Y0)
MIN_PROBE_EDGE = 6 <(arbitrary, larger than 6 to avoid having to clamp in this case)
```

```
LEFT =   BED_CORNER_OFFSET_X + MIN_PROBE_EDGE
RIGHT =  BED_CORNER_OFFSET_X - MIN_PROBE_EDGE + BED_SIZE_X
FRONT =  BED_CORNER_OFFSET_Y + MIN_PROBE_EDGE
BACK =   BED_CORNER_OFFSET_Y - MIN_PROBE_EDGE + BED_SIZE_Y
```
would result in 

```
LEFT = 18
RIGHT = 312
FRONT = 0
BACK = 293
```

which is a perfectly symmetrical, 6mm inset probing area across my bed. 

Of course these calculations would need to 
- clamp in order to avoid hitting the limit switches (like in my case, if the inset were anything lower than 6)
- take the nozzle-probe offset into account

but I omitted them here for simplicity.

**Comment by fiveangle** [comment]
Good discussion. But this is an incremental improvement to an otherwise tedious manual configuration that new users should be omitted from being required to learn about. It far exceeds the 80/20 rule for user coverage and should be the default out of the box in my opinion, else it should not be implemented it at all since it would not serve its intended purpose to ease bring up for the masses.

Anyone who upgrades Marlin from their previous configuration must change these values accordingly, whether by adjusting sample integers or in addition, removing two "/" chars, there is no perceivable impact to them. The benefit to new users does not outweigh the near-zero inconvenience to the existing user base IMO.

**Comment by thinkyhead** [comment]
>
```
X_MIN_POS
X_MAX_POS
Y_MIN_POS
Y_MAX_POS
```
> If those are set, wouldn't any calulation of those be good? Even asymmetrical beds?

Those specify the physical limits of the axes, and are not in any way related to the bed size. The bed can't be assumed to be exactly centered between them.

**Comment by thinkyhead** [comment]
> Do we have bed size defined anywhere explicitly, or is it inferred from the travel limits?

That is the one missing piece of data that would make automation possible.

File: Marlin/Conditionals_post.h [hunk 2]
```diff
@@ -708,6 +708,22 @@
   #endif
 
   /**
+   * Set the boundaries for probing (where the probe can reach) if user has not already configured.
+   */
+  #ifndef LEFT_PROBE_BED_POSITION
+    #define LEFT_PROBE_BED_POSITION ((X_PROBE_OFFSET_FROM_EXTRUDER) <= (MIN_PROBE_EDGE) ? (X_MIN_POS + MIN_PROBE_EDGE) : (X_MIN_POS + X_PROBE_OFFSET_FROM_EXTRUDER))
```
Comment (thinkyhead) [inline Marlin/Conditionals_post.h:714]: `X_MIN_POS` is not related to the position of the bed, so this only adds to the confusion. Also, these bounds only apply to `LINEAR` and `BILINEAR` leveling.

**Comment by thinkyhead** [comment]
So…  I think this PR only continues to exacerbate the original problem, which is that we do not track the bed boundaries, and therefore can't guarantee good values. I would propose starting by adding the bed boundaries as new options, and then build from there. That will also provide more intelligent behavior in other places.

Note that you can use `static_assert` to do things that `#if` can't do.

Note also that there are legitimate cases where you want to have your probing grid points set up symmetrically on the bed even though some of them cannot be reached and probed. The un-probed points can be filled in by extrapolating from the nearby probed points.

**Comment by fiveangle** [comment]
Data not available for 100% of supported devices, so I guess we can't improve it without deeper changes.

**Comment by thinkyhead** [comment]
The MIN_POS for most (non-Delta) beds can be safely assumed to be zero. The bed size can be estimated as the nearest round number smaller than the physical bounds. For example, from the Anet A8 config…
```cpp
#define X_MIN_POS -33
#define Y_MIN_POS -10
#define Z_MIN_POS 0
#define X_MAX_POS 220
#define Y_MAX_POS 220
#define Z_MAX_POS 240
```
…we can estimate bed bounds as…
```cpp
#define BED_X_MIN 0
#define BED_Y_MIN 0
#define BED_X_MAX 200
#define BED_Y_MAX 200
```
…or, assuming all beds are centered or have their limit at 0,0…
```cpp
#define BED_SIZE_X 200
#define BED_SIZE_Y 200
```

File: Marlin/Conditionals_post.h (hunk 2 above)
Comment (fiveangle) [reply Marlin/Conditionals_post.h:714]: That is all they were intended to address.

**Comment by fiveangle** [comment]
With #7341 this can be revisited

**Pull request closed without merging**